---
title: "Slide Seq Differential expression"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{slideseqde}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Imports

```{r setup}
library(slideseqde)
library(dplyr)
library(ggplot2)
library(RCTD)
library(Matrix)
```

# RCTD to deconvolv and transfer annotation
## Functions to prepare reference for RCTD

Prepare Reference
```{r}
# refdir = '~/Workplace/slide-seq-de/shenglin_raw_data/scRNA.refernece'
make.RCTD.reference = function(refdir){
  dge_file = file.path(refdir, "dge.rds")
  raw.data= readRDS(dge_file)
  
  meta_data = read.csv(file.path(refdir,"meta_data.csv"))
  rownames(meta_data) = meta_data$barcode
  
  common_barcodes = intersect(colnames(raw.data), rownames(meta_data))
  raw.data = raw.data[, common_barcodes]
  meta_data = meta_data[common_barcodes, ]
  
  cell_dict = read.csv(paste(refdir, "cell_type_dict.csv", sep = "/"))
  colnames(cell_dict) = c('cluster','cluster_name')
  meta_data = inner_join(meta_data, cell_dict)
  cell_types = meta_data$cluster_name
  names(cell_types) = meta_data$barcode
  
  cell_types = as.factor(cell_types)
  nUMI = meta_data$nUMI
  names(nUMI) = meta_data$barcode
  
  reference <- Reference(raw.data, cell_types, nUMI)
  reference
}

```


## Make the rctd reference
```{r}
rctd.reference =  make.RCTD.reference('~/Workplace/slide-seq-de/shenglin_raw_data/scRNA.refernece')
```
## Make the rctd  Epithelial reference
```{r}
rctd.epi.reference =  make.RCTD.reference('~/Workplace/slide-seq-de/shenglin_raw_data/epi/ano.epi')
```



## Load Spatial Data
```{r}
dnl <- c(
    'Tumor01' = '~/Workplace/slide-seq-de/shenglin_raw_data/2020-11-25_Puck_201014_23/hg19.exonic+intronic/',# P3
    'Tumor02' = '~/Workplace/slide-seq-de/shenglin_raw_data/2020-11-25_Puck_201014_24/hg19.exonic+intronic/'# P3
)
cdl = lapply(dnl[c('Tumor01', 'Tumor02')],createSparseSpatialObjFromRCTD)
```


## Make RCTD spatial dataset
```{r}
make.RCTD.spatial = function(spatial.obj, selected.barcodes = NULL){
  counts = as.matrix(spatial.obj$cm)
  coords = spatial.obj$pos
  rownames(coords) = coords$barcodes
  coords$barcodes = NULL
  if(!is.null(selected.barcodes)){
    counts = counts[,selected.barcodes]
    coords = coords[selected.barcodes,]
  }
  nUMI <- colSums(counts)
  puck <- RCTD::SpatialRNA(coords, counts, nUMI)
  puck
}
```

## SpatialRNA call
```{r}
RCTD.pucks = lapply(cdl, make.RCTD.spatial)
```

```{r}
myRCTD <- lapply(RCTD.pucks, function(puck) {create.RCTD(puck, rctd.reference, max_cores = 10)})
```

```{r}
myRCTD <- lapply(myRCTD, function(puck) {run.RCTD(puck, doublet_mode = 'doublet')})
```

Store RCTD 
```{r}
saveRDS(myRCTD, '~/Workplace/slide-seq-de/shenglin_raw_data/my_RCTD.rds')
```

Results from RCTD
```{r}
results = myRCTD$Tumor01@results
norm_weights = sweep(results$weights, 1, Matrix::rowSums(results$weights), '/')
cell_type_names <- myRCTD$Tumor01@cell_type_info$info[[2]] #list of cell type names
spatialRNA <- myRCTD$Tumor01@spatialRNA
```
Find out doublets
```{r}
doublets <- results$results_df[results$results_df$spot_class == "doublet_certain",]
doub_occur <- table(doublets$second_type, doublets$first_type)
```
Second level Annotation
Take out the `Epithelial Cells`
Certain Epithelial cells and singlets
```{r}
RCTD.epi.pucks = lapply(c('Tumor01','Tumor02'), function(puck.name)  {
  results = myRCTD[[puck.name]]@results
  epi.barcodes = results$results_df %>% dplyr::filter(first_type == 'Epitheial') %>% rownames
  spatial.obj = cdl[[puck.name]]
  make.RCTD.spatial(spatial.obj, epi.barcodes)
})
```
Second level Run 
```{r}
epi.myRCTD <- lapply(RCTD.epi.pucks, function(puck) {create.RCTD(puck, rctd.epi.reference, max_cores = 10)})
epi.myRCTD <- lapply(epi.myRCTD, function(puck) {run.RCTD(puck, doublet_mode = 'doublet')})
```
```{r}
names(epi.myRCTD) = c('Tumor01','Tumor02')
```

# Create annotation for spatial data

Add this annotation to the existing annotation 
```{r}
annotation.RCTD = lapply(c('Tumor01','Tumor02'), function(puck.name){
  prev.res = myRCTD[[puck.name]]@results$results_df %>%  dplyr::filter(first_type != 'Epitheial')
  new_res = rbind(prev.res, epi.myRCTD[[puck.name]]@results$results_df )
  new_res = new_res %>% dplyr::filter((spot_class == 'singlet') || (spot_class == 'doublet_certain'))
  anno = as.factor(new_res$first_type)
  names(anno) = paste(puck.name ,rownames(new_res),sep = '_')
  anno
})
#names(annotation.RCTD) = c('Tumor01','Tumor02')
```

# Process spatial data
## Covert the Spatial RNA-seq to spatial object

Covert to pagoda objects
```{r}
objs = convertSpatialObj2Pagoda(cdl)
```

```{r}
# anno <- readRDS('~/Workplace/slide-seq-de/data/annotation/slide.seq.ano.v2.rds')
```

## Assign RCTD annotations
```{r}
annotation.RCTD.all = do.call(c, annotation.RCTD)
saveRDS(annotation.RCTD.all, '~/Workplace/slide-seq-de/data/annotation/annotation_RCTD_all.rds')
objs = updateSpatialInfoCustom(objs, annotation.RCTD.all)
#objs = updateSpatialInfo(objs, anno)
``` 

## Add KDE boundary around Tumor cell type to create
```{r}
objs$cdl.p2$Tumor01 = createKDEBoundary(objs$cdl.p2$Tumor01)
objs$cdl.p2$Tumor02 = createKDEBoundary(objs$cdl.p2$Tumor02)
```

Compute joint context
```{r}
# objs$cdl.p2$Tumor01 = computeJointContext(
#     objs$cdl.p2$Tumor01,
#     annotation.type.name = "annot.L2",
#     joint.type.name = "jcontext.L2"
# )
# 
# objs$cdl.p2$Tumor02 = computeJointContext(
#     objs$cdl.p2$Tumor02,
#     annotation.type.name = "annot.L2",
#     joint.type.name = "jcontext.L2"
# )

objs$cdl.p2$Tumor01 = computeJointContext(objs$cdl.p2$Tumor01)

objs$cdl.p2$Tumor02 = computeJointContext(objs$cdl.p2$Tumor02)
```

Compute marker genes
```{r}
marker.de <- lapply(objs$cdl.p2, function(p2){
    p2$getDifferentialGenes(
    type="PCA",clusterType="annot",verbose=T,append.auc=T, upregulated.only=T)
})
```
## Get tumor markers
```{r}
tumor.markers.01 = marker.de[['Tumor01']][['Tumor']] %>% arrange(-AUC) %>% rownames
tumor.markers.02 = marker.de[['Tumor02']][['Tumor']] %>% arrange(-AUC) %>% rownames
tumor.markers <- intersect(
    marker.de[['Tumor01']][['Tumor']] %>% arrange(-AUC) %>% rownames,
    marker.de[['Tumor02']][['Tumor']] %>% arrange(-AUC) %>% rownames
)
```



## Keep extra genes
```{r}
angiogenesis.genes <- readLines(con <- file(
    "/home/hsarkar/Workplace/slide-seq-de/data/de_results/angiogenesis.list"))

emt.genes <- readLines(con <- file(
    "/home/hsarkar/Workplace/slide-seq-de/data/de_results/emt.list"))

lr.pairs = read.csv('/d0/home/meisl/Workplace/Prostate/healty.data/conos/F7.slide.seq/all.sig.csv')
lr.genes = unique(union(lr.pairs$receptor , lr.pairs$ligand))


extra.genes <- union(angiogenesis.genes, emt.genes)
extra.genes <- union(extra.genes, lr.genes)
```

We want to subset the cells according to the distance from the boundary
```{r}
# Tumor01
distance.subsets.01 = get.inner.embedding(objs$cdl.p2$Tumor01, 
                                       tumor.center = c(4250, 3250),
                                       nontumor.center = c(2000, 2500)
                                    )

distance.subsets.02 = get.inner.embedding(objs$cdl.p2$Tumor02, 
                                       tumor.center = c(2500, 2700),
                                       nontumor.center = c(4250, 3250)
                                    )
```

Create a distance
Plot the with respect to a particular distance from the boundary
```{r,fig.width=10, fig.height=4}
distance.from.boundary = 350
p1 = ggplot2::ggplot(
        distance.subsets.01 %>% dplyr::filter(mindist > distance.from.boundary ),
        aes(x=xcoord, y=ycoord, color=mindist)
      ) +
      ggtitle('Tumor01') +
      ggplot2::geom_point(size=0.3)+
      objs$cdl.p2$Tumor01$misc$cont

p2 = ggplot2::ggplot(
        distance.subsets.02 %>% dplyr::filter(mindist > distance.from.boundary ),
        aes(x=xcoord, y=ycoord, color=mindist)
      ) +
      ggtitle('Tumor02') +
      ggplot2::geom_point(size=0.3)+
      objs$cdl.p2$Tumor02$misc$cont

p1 | p2
```


If this distance is what we want, subset cells with to remove the cells near the boundary we will make corrected profiles with this distance,
```{r}
profile.list = list()
distance.from.boundary=350
puck.profiles = lapply(c('Tumor01','Tumor02'),function(puck.name){
    t1 <- objs$cdl.p2[[puck.name]]
    t1.raw <- objs$cdl[[puck.name]]
    
    if(puck.name == "Tumor01"){
      tumor.center = c(4250, 3250)
      nontumor.center = c(2000, 2500)
    }else{
      tumor.center <- c(2500, 2700)
      nontumor.center <- c(4250, 3250)
    }
    
    slideseqde::preparePseudoBulkProfile(
        t1,
        t1.raw,
        distance.from.boundary, 
        tumor.center,
        nontumor.center,
        extra.genes
    )
})
names(puck.profiles) = c('Tumor01','Tumor02')
profile.list[[1]] = as.matrix(puck.profiles[['Tumor01']]$rprofile)
profile.list[[2]] = as.matrix(puck.profiles[['Tumor02']]$rprofile)
```



# Computing the DE for different cell types
## Use BVLS
```{r}
target_ctypes = c(
    'Pericytes',
    'Fibroblasts',
    'Endothelial',
    'Macrophage'   
)

multi.de.res =  lapply(target_ctypes, function(target){
        print(target)
        target1 = paste(target, 'Tumor', sep='_')
        target2 = paste(target, 'Non_Tumor', sep='_')
        de.res =  slideseqde::getDifferentialExpressionBVLS(profile.list, target1, target2,
                                 de.method = 'edgeR.lrt')
        if(is.null(de.res)){
          return(NULL)
        }
        de.plot = scatterPlotWithCorrection(de.res, intersect(tumor.markers.01,
                                                        tumor.markers.02)[1:50],
                                                       target,20)
        # rownames(pvals.filtered) = 1:nrow(pvals.filtered)
        list(de.res = de.res, 
             scatter.plot = de.plot
        )
      }
)
names(multi.de.res) = target_ctypes
```

# Plot the DE correction the "Fibroblasts"
```{r}
multi.de.res$Fibroblasts$scatter.plot
```

